test/suites/drift/suite_test.go

```diff
@@ -57,71 +69,660 @@ var _ = AfterEach(func() { env.Cleanup() })
+			// The test is still written to expect maximum of 2 nodes disruped/deleted at a time.
+			nodePool.Spec.Disruption.Budgets = []karpv1.Budget{{
+				Nodes: "50%",
+			}}
+			var numPods int32 = 6
+			dep = coretest.Deployment(coretest.DeploymentOptions{
+				Replicas: numPods,
+				PodOptions: coretest.PodOptions{
+					ObjectMeta: metav1.ObjectMeta{
+						Annotations: map[string]string{
+							karpv1.DoNotDisruptAnnotationKey: "true",
+						},
+						Labels: map[string]string{"app": "large-app"},
+					},
+					// Each Standard_DS4_v2 has 8 cpu, so each node should fit 2 pods.
+					ResourceRequirements: corev1.ResourceRequirements{
+						Requests: corev1.ResourceList{
+							corev1.ResourceCPU: resource.MustParse("3"),
+						},
+					},
+				},
+			})
+			selector = labels.SelectorFromSet(dep.Spec.Selector.MatchLabels)
+			env.ExpectCreated(nodeClass, nodePool, dep)
 
-	// TODO: Add budget tests
+			nodeClaims := env.EventuallyExpectCreatedNodeClaimCount("==", 3)
+			nodes := env.EventuallyExpectCreatedNodeCount("==", 3)
+			env.EventuallyExpectHealthyPodCount(selector, int(numPods))
 
-	It("should deprovision nodes that have drifted due to labels", func() {
+			// List nodes so that we get any updated information on the nodes. If we don't
+			// we have the potential to over-write any changes Karpenter makes to the nodes.
+			// Add a finalizer to each node so that we can stop termination disruptions
+			By("adding finalizers to the nodes to prevent termination")
+			for _, node := range nodes {
+				Expect(env.Client.Get(env.Context, client.ObjectKeyFromObject(node), node)).To(Succeed())
+				node.Finalizers = append(node.Finalizers, common.TestingFinalizer)
+				env.ExpectUpdated(node)
+			}
 
-		By(fmt.Sprintf("creating pod %s, nodepool %s, and nodeclass %s", pod.Name, nodePool.Name, nodeClass.Name))
-		env.ExpectCreated(pod, nodeClass, nodePool)
+			By("making the nodes empty")
+			// Delete the deployment to make all nodes empty.
+			env.ExpectDeleted(dep)
 
-		By(fmt.Sprintf("expect pod %s to be healthy", pod.Name))
-		env.EventuallyExpectHealthy(pod)
+			// Drift the nodeclaims
+			By("drift the nodeclaims")
+			nodePool.Spec.Template.Annotations = map[string]string{"test": "annotation"}
+			env.ExpectUpdated(nodePool)
 
-		By("expect created node count to be 1")
-		env.ExpectCreatedNodeCount("==", 1)
+			env.EventuallyExpectDrifted(nodeClaims...)
+
+			env.ConsistentlyExpectDisruptionsUntilNoneLeft(3, 2, 5*time.Minute)
+		})
+		It("should respect budgets for non-empty delete drift", func() {
+			nodePool = coretest.ReplaceRequirements(nodePool,
+				karpv1.NodeSelectorRequirementWithMinValues{
+					NodeSelectorRequirement: corev1.NodeSelectorRequirement{
+						Key:      corev1.LabelInstanceTypeStable,
+						Operator: corev1.NodeSelectorOpIn,
+						Values:   []string{"Standard_DS4_v2"},
+					},
+				},
+			)
+			// We're expecting to create 3 nodes, so we'll expect to see at most 2 nodes deleting at one time.
```

Comment:
Duplicate of the same comment around async drift not being supported. However it does say `at most`, so maybe just ignore here?

I mean the plan is the support will eventually be there.

Reply:
Updated comment